File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        answer = 0
        val_max = prices[0]
        val_min = prices[0]
        for i in range(len(prices) - 1):
            if prices[i] < prices[i + 1]:
                val_max = max(val_max, prices[i + 1])
                answer = max(answer, val_max - val_min)
            if prices[i] > prices[i + 1] and prices[i + 1] < val_min:
                val_max = prices[i + 1]
                val_min = prices[i + 1]
        return answer

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.dart
class Solution {
  int maxProfit(List<int> prices) {
    int max_p = prices[0];
    int min_p = prices[0];
    int answer = 0;
    for(int i = 0;i < prices.length-1;i++){
        if (prices[i] > prices[i + 1] && prices[i + 1] < min_p) {
            min_p = prices[i + 1];
            max_p = prices[i + 1];
        }
        if (prices[i] < prices[i+1]) {
            max_p = prices[i + 1];
            answer = max(answer, max_p - min_p);
        }
    }
    return answer;
  }
}